base/floatfuncs.jl

```diff
@@ -20,6 +20,7 @@ maxintfloat(::Type{Float64}) = 9007199254740992.
 maxintfloat(::Type{Float32}) = Float32(16777216.)
 maxintfloat(::Type{Float16}) = Float16(2048f0)
 maxintfloat(x::T) where {T<:AbstractFloat} = maxintfloat(T)
+maxintfloat(::Type{S}, ::Type{T}) where {S<:AbstractFloat, T<:Integer} = (m = maxintfloat(S); Int === Int32 ? min(m, S(typemax(T))) : m)
```

Comment:
So, if someone types `maxintfloat(Float64, Int16)` then I think they'll say this is producing the wrong answer.

The good news is you don't have to work so hard, because LLVM takes care of optimizing this for you:
```julia
julia> mymaxintfloat(::Type{S}, ::Type{T}) where {S<:AbstractFloat, T<:Integer} = min(maxintfloat(S), S(typemax(T)))
mymaxintfloat (generic function with 1 method)                                                                                                                
                                                                                                                                                              
julia> mymaxintfloat(Float64, Int32)
2.147483647e9                                                                                                                                                 
                                                                                                                                                              
julia> mymaxintfloat(Float64, Int)
9.007199254740992e15

julia> @code_llvm mymaxintfloat(Float64, Int32)

define double @julia_mymaxintfloat_60880(i8**, i8**) #0 !dbg !5 {
top:
  ret double 0x41DFFFFFFFC00000
}

julia> mymaxintfloat(Float64, Int16)
32767.0

julia> @code_llvm mymaxintfloat(Float64, Int16)

define double @julia_mymaxintfloat_60922(i8**, i8**) #0 !dbg !5 {
top:
  ret double 3.276700e+04
}
```